Include/Rover/Concept.hpp

```diff
@@ -0,0 +1,28 @@
+#ifndef CONCEPT_HPP
```

Comment:
My concern about that is described in my comment on Constant.hpp file.
While I personally do not find this way to emulate concepts very useful, I need to know how to address that issue before I can proceed with removal.

Reply:
A ```Constant<T>``` where ```T``` is itself a generator is a valid ```Constant```. It might be useful in creating higher order generators, it might not be, but there's no more reason to assert that ```T``` is not a generator than there is to assert that ```T``` is an ```std::mutex``` or some other unusual type.

Reply:
That is correct, but we need to think how to make the user syntax clear. I mean, the very reason there was such an assertion was that I came across that issue while testing - and I would stumble upon it a few more times.
If I disable construction of non-reference generators, it should much safer and easier to augment.

Reply:
Okay, I might need to rethink the API for this then. Reference's are not supposed to be a big part of this library and their use case is very limited to circumstances where some parameters have a logical dependency on other parameters. The overwhelming majority of parameters have no dependencies and don't need any sort of reference semantics. I do not want this minor use case to have be a major part of how generators work.

I'll take a day or so to think about it and we can discuss it tomorrow morning.

Reply:
Independent References behave in the same way as independent Generators, but References handle dependencies of any kind. The memory overhead is between 3 * sizeof(size_t) and 4 * sizeof(size_t) to accommodate a reference counter (can be further lowered by removing the weak counter during the course of shared_optional implementation), which is negligible for any use case. Session traversal imposes much greater overhead, yet still negligible for any real use case with say < 1K parameters on virtually any computer device. Even using a pointer, dynamic memory allocation is made only once, when a Reference is created, so I would not even optimizing those out by implementing shared_optional.

Most parameters may be independent, but the idea is when there is some sort of copying or assignment, it far more likely implies referencing rather than a hard copy. It is especially true for generators like Range and arithmetic ones, such as Range(a, 10) and (a + 10). Especially so in contexts where referencing is the default behavior, such as Python. If one is used to working with Generators directly and has to wrap each Generator to a special class whenever they want to define a dependency, it is like asking for a mistake.

The entire library does not have to depend on how References work as long as all free Generators are References. That is why they were originally called Parameters. The generate() call will still pass through reference functionality all the time, right, but it is in the same way every Range uses std::random etc. A person implementing a Generator does not need to deal with, or even be aware of, that implementation details.

Sure, let me know.

Reply:
Okay I still need to think about it but I think it's clear that it won't be possible to statically/compile time generate a Session. It will have to be dynamically generated and use type erasure to allow for heterogeneous types.

Reply:
Why? It does work in this way, right now, with the static Session and without type erasure. The only difference that I allow creating free generators, which I can simply prohibit.